import-meta-path-helpers.md

```diff
@@ -0,0 +1,76 @@
+# Helpers on `import.meta` for working with paths
+
+## Background
+
+A common user request from developers writing ESM code is easy access to file path information of the current module. Or put another way, they miss the `__dirname` and `__filename` local variables from CommonJS, that define the directory and full path of the current module. These values are distinct from `import.meta.url`, which is the URL of the current module. To illustrate:
+
+- `import.meta.url` is a string like `file:///Users/Geoffrey/my-project/src/index.js`
+- `__filename` is a string like `/Users/Geoffrey/my-project/src/index.js`
+- `__dirname` is a string like `/Users/Geoffrey/my-project/src`
+
+Since `__filename` and `__dirname` have been around [since Node.js 0.1.27](https://nodejs.org/api/modules.html#__dirname), released in 2010, they are very well known to Node.js developers and very broadly referenced by `npm` packages. It is common to find documentation suggesting passing `__dirname` into a function for serving the static assets of a folder, for example, or passing `__filename` into a function used for logging.
+
+In Node.js, developers working with such path-based APIs need to jump through a few hoops. Node.js provides a helper `fileURLToPath` that developers can import from `node:url`, to enable code like `serveStaticAssets(fileURLToPath(dirname(import.meta.url)))`. A common complaint of Node.js developers is that this is too verbose, and a way to achieve the same with simpler code and without needing to import utility helpers would be preferable.
+
+## Proposal
+
+For runtimes that support filesystem-based modules, we propose adding two new optional properties to `import.meta`:
+
+### `import.meta.filename`
+
+A string containing the full absolute filesystem path to the current module, like CommonJS `__filename`. This is the same as `fileURLToPath(import.meta.url)` in Node.js. An example value would be `/Users/Geoffrey/my-project/src/index.js`.
+
+The value will be undefined for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
+
+### `import.meta.dirname`
+
+A string containing the full absolute filesystem path of the folder containing the current module, like CommonJS `__dirname`. This is the same as `dirname(fileURLToPath(import.meta.url))` in Node.js. An example value would be `/Users/Geoffrey/my-project/src`.
+
+The value should not have a trailing slash. This is consistent with CommonJS `__dirname`, so that it can be used as a drop-in replacement.
```

Comment:
It’s also consistent with filesystems, where the trailing slash means inside the directory, and without the slash means the directory.